janino/src/main/java/org/codehaus/janino/UnitCompiler.java

```diff
@@ -9324,35 +9326,35 @@
 
         if (maximallySpecificIInvocables.size() == 1) return (IInvocable) maximallySpecificIInvocables.get(0);
 
-        ONE_NON_ABSTRACT_INVOCABLE:
+        ONE_CONCRETE_INVOCABLE:
         if (maximallySpecificIInvocables.size() > 1 && iInvocables[0] instanceof IClass.IMethod) {
 
             // Check if all methods have the same signature (i.e. the types of all their parameters are identical) and
-            // exactly one of the methods is non-abstract (JLS7 15.12.2.2.BL2.B1).
-            IClass.IMethod theNonAbstractMethod = null;
+            // exactly one of the methods is concrete (JLS8 15.12.2.5.B8.B1).
+            IClass.IMethod theConcreteMethod = null;
             {
                 Iterator<IClass.IInvocable> it                          = maximallySpecificIInvocables.iterator();
                 IClass.IMethod              m                           = (IClass.IMethod) it.next();
                 final IClass[]              parameterTypesOfFirstMethod = m.getParameterTypes();
                 for (;;) {
-                    if (!m.isAbstract()) {
-                        if (theNonAbstractMethod == null) {
-                            theNonAbstractMethod = m;
+                    if (!m.isAbstract() && !m.isDefault()) {
+                        if (theConcreteMethod == null) {
+                            theConcreteMethod = m;
                         } else {
                             IClass declaringIClass                     = m.getDeclaringIClass();
-                            IClass theNonAbstractMethodDeclaringIClass = theNonAbstractMethod.getDeclaringIClass();
+                            IClass theNonAbstractMethodDeclaringIClass = theConcreteMethod.getDeclaringIClass();
                             if (declaringIClass == theNonAbstractMethodDeclaringIClass) {
-                                if (m.getReturnType() == theNonAbstractMethod.getReturnType()) {
+                                if (m.getReturnType() == theConcreteMethod.getReturnType()) {

```

Comment:
Perhaps add `&& !(m.getDeclaringClass().isInterface() && theConcreteMethod.getDeclaringClass().isInterface())`

To implement:
> Otherwise, if all the maximally specific methods are abstract or default 
[we know here that they are not abstract, so they must be defined in an interface],
> and the signatures of all of the maximally specific methods have the same erasure (§4.6), 
> then the most specific method is chosen arbitrarily among the subset of the maximally 
> specific methods that have the most specific return type.

Reply:
(Note that you'd also have to add the check that they have the same erasure.)

Reply:
In this loop, won't all the `getDeclaredIClass` of the concrete methods it examines all not be interfaces? methods in interfaces would be abstract or default, and both are excluded. 

Or are you saying check this _instead of_ defining `isDefault`?

I'm new to the janino APIs but it looks like `IClass` does not carry info about generic types; they're erased already. If that's so, then I think the code is already comparing erasures?